### PR TITLE
cubeHelix interpolation on hue space

### PR DESCRIPTION
**Pull request by dangoldbj**

### CODE REVIEW

File: src/lib-components/colors.ts [hunk 1]
```diff
@@ -36,26 +46,42 @@ export function hslToHex(hParam: number, sParam: number, lParam: number): string
...
+export function cubeHelix(h: number, s: number, l: number) {
+  const A = -0.14861;
+  const B = +1.78277;
+  const C = -0.29227;
+  const D = -0.90649;
+  const E = +1.97294;
+  const hRad = (h + 120) * (Math.PI / 180);
+  const a = s * l * (1 - l);
+  const cosh = Math.cos(hRad);
+  const sinh = Math.sin(hRad);
```
Comment (stared) [inline src/lib-components/colors.ts]: `sinh` and `cosh` usually mean hyperbolic functions <https://en.wikipedia.org/wiki/Hyperbolic_functions>, but OK :)

File: src/lib-components/colors.ts [hunk 2]
```diff
@@ -36,26 +46,42 @@ export function hslToHex(hParam: number, sParam: number, lParam: number): string
...
  * https://github.com/stared/quantum-game/blob/master/js/transition_heatmap.js
  */
 export function colorComplex(re: number, im: number): string {
   const angleInDegrees = ((Math.atan2(im, re) * 360) / TAU + 360) % 360;
-  const r = Math.sqrt(re * re + im * im); // for pure color it should be always 1
-  return hslToHex(angleInDegrees, 100, 100 - 50 * r);
+  const radius = Math.sqrt(re * re + im * im); // for pure color it should be always 1
+  const [r, g, b] = cubeHelix(angleInDegrees, 1, 1 - 0.5 * radius);
+  return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
 }
```
Comment (stared) [inline src/lib-components/colors.ts]: Here I would expect *a few* functions, or maybe `complexColors` with a parameter. To add new color schemes and support multiple, not to *change* the previous one.

File: src/lib-components/colors.ts [hunk 3]
```diff
@@ -36,26 +46,42 @@ export function hslToHex(hParam: number, sParam: number, lParam: number): string
...
 
 export function colorComplexPhaseToHue(re: number, im: number, saturation = 100, lightness = 50): string {
   const angleInDegrees = ((Math.atan2(im, re) * 360) / TAU + 360) % 360;
-  return hslToHex(angleInDegrees, saturation, lightness);
+  const [r, g, b] = cubeHelix(angleInDegrees, saturation / 100, lightness / 100);
+  return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
```
Comment (stared) [inline src/lib-components/colors.ts]: Likewise above

File: src/lib-components/complex-legend.vue [hunk 4]
```diff
@@ -191,7 +191,7 @@ export default defineComponent({
     fill: rgba(0, 0, 0, 0.3);
   }
   & .phase-arc {
-    stroke: rgba(0, 0, 0, 0.1);
+    stroke: rgba(0, 0, 0, 0.5);
```
Comment (stared) [inline src/lib-components/complex-legend.vue:194]: I preffered the previous option

File: src/lib-components/colors.ts (hunk 1 above)
Comment (Frizi) [reply src/lib-components/colors.ts]: possible replacement: `sinHue`, `cosHue` if you prefer that.

File: src/lib-components/colors.ts (hunk 2 above)
Comment (Frizi) [reply src/lib-components/colors.ts]: I don't think there is much to abstract here. Each color space convertion function will have its own distinct inputs anyway, and you will need to convert complex number into them in some way. That mapping depends on specific function. Making this more generic doesn't really save you anything - you need both parts to be color space specific. The only "generic" thing left is text formatting, which can be separate.

I guess taking arg/norm of complex should also be a function call, but it is still just an implementation detail. You might not need them in some situations.

I think a better idea would be to have a standardized "complex to color" function signature that takes `Complex` and returns `[r,g,b]`. That way we can substitute those functions to choose any palette you want, or even allow users to implement their own.

File: src/lib-components/colors.ts (hunk 3 above)
Comment (Frizi) [reply src/lib-components/colors.ts]: Additionaly, lightness can for sure be higher. Hsl keeps the "most saturation" point at 50% lightness, so it was the previous default. At 100% lightness it always returned white. I believe cubeHelix doesn't do that. Try giving it maximum allowed value and see what happens.

**Review by Frizi — COMMENTED** [review]
If possible, I would still love to see oklab implementation as additional color scheme, next to hsl and cubeHelix. You can use the exact same approach - Use Lab -> sRGB conversion function, provide it with variable hue to generate rainbow scheme.

[Here's how that looks](https://bottosson.github.io/posts/oklab/#comparing-oklab-to-hsv), and below is an example implementation of that conversion function.

File: src/lib-components/colors.ts (hunk 2 above)
Comment (stared) [reply src/lib-components/colors.ts]: @Frizi The issue is different - the old one was not kept, but replaced. 
We want to be able to write a few functions.

**Comment by Frizi** [comment]
Fixed Oklab color space, found nicely looking parameters for it. Added debug visualizer for color palettes.
image(1)
image(2)

**Review by Frizi — APPROVED** [review]

File: src/lib-components/complex-legend.vue (hunk 4 above)
Comment (dangoldbj) [reply src/lib-components/complex-legend.vue:194]: The previous option doesn't work well with the oklab implementaion.

Previous option


Current option

**Pull request merged**